### PR TITLE
Metachain id creation

### LINKED ISSUE #76: Change chainId to metachainId.
- logic for metachainId creation is metachainId = hash(0x19 0x4d domainseparator typehash)
- 0x4d = “M” for Mosaic
- domainseparator = name:“Mosaic-Metachain”, version, Consensus: <Consensus>
- typeHash “MetachainId(address anchor)”<anchor>

### PR DESCRIPTION
**Pull request by gulshanvasnani**
PR contains below changes :
- [x] Creation of metachain id.
- [x] Update usage of chainid to metachain id.

fixes #76

### CODE REVIEW

File: contracts/consensus/Consensus.sol [hunk 1]
```diff
@@ -25,8 +25,9 @@ import "../core/CoreI.sol";
 import "../core/CoreStatusEnum.sol";
 import "../reputation/ReputationI.sol";
 import "../proxies/MasterCopyNonUpgradable.sol";
+import "../version/MosaicVersion.sol";
 
-contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
+contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI, MosaicVersion {
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: move this second

File: contracts/consensus/Consensus.sol [hunk 2]
```diff
@@ -53,10 +54,22 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
...
+    string public constant MOSAIC_DOMAIN_SEPARATOR_NAME = "Mosaic";
+
+    /** It is domain separator typehash used to calculate metachain id. */
+    bytes32 public constant MOSAIC_DOMAIN_SEPARATOR_TYPEHASH = keccak256(
+        "MosaicDomain(string name,string version,uint256 originChainId,address consensus)"
+    );
+
+    /** It is metachain id typehash used to calculate metachain id. */
+    bytes32 public constant METACHAIN_ID_TYPEHASH = keccak256(
+        "Metachain(address anchor)"
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: "MetachainId(address anchor)"

File: contracts/consensus/Consensus.sol [hunk 3]
```diff
@@ -110,15 +123,18 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
     mapping(address => address) public committees;
 
 // NOTE: consider either storing a linked list; or getting rid of it
```
Comment (benjaminbollen) [inline contracts/consensus/Consensus.sol]: you can remove this note pls; now the mapping for `anchors` is more sensible and we'll keep it as this

**Review by benjaminbollen — DISMISSED** [review]
LGTM!

File: contracts/consensus/Consensus.sol [hunk 4]
```diff
@@ -621,6 +647,53 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
...
+     *         0x19 signed data as per EIP-191.
+     *         0x4d is version byte for Mosaic.
+     *         <mosaic-domain-separator> format is `MosaicDomain(string name,
+     *                            string version,uint256 originChainId,
+     *                            address consensus)`.
+     *         <metachainid-typehash> format is MetachainId(address anchor).
+     *
+     *         <mosaic-domain-separator> and <metachainid-typehash> is EIP-712
+     *         complaint.
+     * @param _anchor Anchor address of the new meta-chain.
```
Comment (jayeshbairagi) [inline contracts/consensus/Consensus.sol]: Typo `meta-chain` should be `metachain`

File: contracts/consensus/Consensus.sol [hunk 5]
```diff
@@ -388,7 +414,7 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
      *            core's precommit.
      *          - anchor contract for the given chain id should exist
      *
-     * @param _chainId Chain id.
+     * @param _metachainId metachain id.
```
Comment (jayeshbairagi) [inline contracts/consensus/Consensus.sol]: Typo `metachain` should be `Metachain`

File: contracts/consensus/Consensus.sol [hunk 6]
```diff
@@ -424,7 +450,7 @@ contract Consensus is MasterCopyNonUpgradable, CoreStatusEnum, ConsensusI {
         );
 
         // Makes sure that assigned core is active.
-        address core = assignments[_chainId];
+        address core = assignments[_metachainId];
         require(
             isCore(core),
             "There is no core for the specified chain id."
```
Comment (jayeshbairagi) [inline contracts/consensus/Consensus.sol]: Typo `chain` should be `metachain`

**Review by jayeshbairagi — CHANGES_REQUESTED** [review]
Minor typos

File: test/core/open_metablock.js [hunk 7]
```diff
@@ -131,7 +131,7 @@ contract('Core::openMetablock', (accounts) => {
 
   beforeEach(async () => {
     config.consensusCoreArgs = {
-      chainId: accountProvider.get(),
+      metachainId: web3.utils.keccak256('metachainid'),
```
Comment (deepesh-kn) [inline test/core/open_metablock.js]: We can get a random hash from utils.

**Review by jayeshbairagi — DISMISSED** [review]
LGTM

**Review by deepesh-kn — CHANGES_REQUESTED** [review]
One change: 
Chain id is still used in `test/core/utils.js` file.

Otherwise looks good. 👍

**Review by jayeshbairagi — APPROVED** [review]

**Review by deepesh-kn — APPROVED** [review]

**Comment by jayeshbairagi** [comment]
Tests are failing right now on Travis due to issue related to Ganache. The issue has been created here: https://github.com/mosaicdao/mosaic-1/issues/102.
All the tests are passing locally.

**Pull request merged**